cpp/src/arrow/dataset/file_csv.cc

```diff
@@ -0,0 +1,99 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/dataset/file_csv.h"
+
+#include <memory>
+#include <string>
+
+#include "arrow/csv/options.h"
+#include "arrow/csv/reader.h"
+#include "arrow/dataset/dataset_internal.h"
+#include "arrow/dataset/file_base.h"
+#include "arrow/dataset/type_fwd.h"
+#include "arrow/dataset/visibility.h"
+#include "arrow/result.h"
+#include "arrow/util/iterator.h"
+
+namespace arrow {
+namespace dataset {
+
+using internal::checked_cast;
+using internal::checked_pointer_cast;
+
+static inline csv::ReadOptions default_read_options() {
+  auto defaults = csv::ReadOptions::Defaults();
+  defaults.use_threads = false;
+  return defaults;
+}
+
+static inline Result<std::shared_ptr<csv::StreamingReader>> OpenReader(
+    const FileSource& source, const CsvFileFormat& format) {
+  ARROW_ASSIGN_OR_RAISE(auto input, source.Open());
+
+  auto maybe_reader = csv::StreamingReader::Make(
+      default_memory_pool(), std::move(input), default_read_options(),
+      format.parse_options, format.convert_options);
+  if (!maybe_reader.ok()) {
+    return maybe_reader.status().WithMessage("Could not open IPC input source '",
+                                             source.path(), "': ", maybe_reader.status());
+  }
+
+  return std::move(maybe_reader).ValueOrDie();
+}
+
+/// \brief A ScanTask backed by an Csv file.
+class CsvScanTask : public ScanTask {
+ public:
+  CsvScanTask(std::shared_ptr<const CsvFileFormat> format, FileSource source,
+              std::shared_ptr<ScanOptions> options, std::shared_ptr<ScanContext> context)
+      : ScanTask(std::move(options), std::move(context)),
+        format_(std::move(format)),
+        source_(std::move(source)) {}
+
+  Result<RecordBatchIterator> Execute() override {
+    ARROW_ASSIGN_OR_RAISE(auto reader, OpenReader(source_, *format_));
```

Comment:
It's missing the column selection optimisation. While it won't save the IO, it'll save the memory and cpu time from elided  conversion.

Reply:
Also, the `ScanOptions` should be used to set the `column_names`, `column_types` and `include_columns` options. Unless you prefer to let the CSV reader infer types and then convert them yourself?

Reply:
Assembling these from `ScanOptions` is definitely a goal; the first pass just defers all this to the format. Some of `ConvertOptions` belong in the format (`null_values`, ..., `strings_can_be_null`, ...) while others such as the ones you note do not. I'll rewrite shortly

Reply:
PTAL. There's a subtlety resulting from the loosely typed nature of CSV: we don't know the types of columns materialized but not projected, which could lead to some very nasty errors in edge cases. For example, if filter were `"a"_ == "b"_` but only column `"c"` were projected then we have no information about the types of fields a, b except that they must be identical. I'm not sure what the correct solution here is

Reply:
That sounds nasty indeed. The datasets layer doesn't know the type of these columns at all?

Reply:
The type of field "a" can't be known without the schema from which it originates. When the filter is applied to a dataset in python or R the dataset's schema will be used to add cast expressions to the filter which will handle the necessary conversions. However in c++ we don't pass the dataset schema at scan time

Reply:
So if e.g. two Parquet files in a dataset have different types for `a` or `b` (which may be a data error), that passes silently?

Reply:
That's indeed a data error. Whether parquet or csv, in C++ this would be caught by the ScannerBuilder::Filter setter (which *does* have access to the dataset's schema), so maybe this edge case isn't interesting after all

Reply:
In database systems, it is required to use a pre-determined schema when reading CSV files. If we want to support schema inference, we could use a part of a file to "sniff" the schema but then assume it going forward. That's a bit better than forcing people to write down the schema

Reply:
Reading using a predetermined schema is already supported for CSV and for all the other formats. The first chunk (as determined by `CsvFileFormat::chunk_size`) is used to sniff the schema which is assumed for the rest of the file, and this inspection can be applied to just one or to all the files in a dataset at discovery time.

Reply:
The edge case here is due to the way we pass only the projected schema during a scan, so maybe the solution is to always pass the full dataset schema as well so that no materialized fields will be inferring at scan time

Reply:
I'll add a failing test tomorrow so this can become less hand wavy

Reply:
https://github.com/apache/arrow/pull/7033/commits/37c493695cb5603b7836291985e6355c3cf1de82

Reply:
Follow up JIRA: https://issues.apache.org/jira/browse/ARROW-8630